PWGHF: Add Ds task and selector

stefanopolitano opened the pull request:
This PR contains the Ds -> phi pi -> K K pi task and selector. 
This PR continues the discussion from [AliceO2Group/O2Phyisics#96](https://github.com/AliceO2Group/O2Physics/pull/96) and[ AliceO2Group/Run3Analysisvalidarion#323](https://github.com/AliceO2Group/Run3Analysisvalidation/pull/323)

The last comments from the previous PR were addressed and the macros were moved to the proper folder (D2H).

fgrosa reviewed (approved):
The change in the event filtering code is trivial, so I approve it already

vkucera commented on PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:
```diff
@@ -0,0 +1,171 @@
+        hfSelDsToKKPiCandidate(statusDsToKKPi, statusDsToPiKK);
+        continue;
+      }
+      SETBIT(statusDsToKKPi, aod::SelectionStep::RecoTopol);
+      SETBIT(statusDsToPiKK, aod::SelectionStep::RecoTopol);
```
You should set these based on the values of `topoDsToKKPi` and `topoDsToPiKK`.

vkucera commented on PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:
```diff
@@ -0,0 +1,171 @@
+               pidTrackNegKaon == TrackSelectorPID::Status::PIDAccepted &&
+               pidTrackPos2Kaon == TrackSelectorPID::Status::PIDAccepted) {
+        SETBIT(statusDsToPiKK, aod::SelectionStep::RecoPID); // accept DsPiKK
+      } else {
+        hfSelDsToKKPiCandidate(statusDsToKKPi, statusDsToPiKK);
+        continue;
+      }
```
Is this part needed?

vkucera commented on PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:
```diff
@@ -0,0 +1,171 @@
+          pidTrackPos2Pion == TrackSelectorPID::Status::PIDAccepted) {
+        SETBIT(statusDsToKKPi, aod::SelectionStep::RecoPID); // accept DsKKPi
+      }
+      // checking PID for Ds to PiKK hypothesis
+      else if (pidTrackPos1Pion == TrackSelectorPID::Status::PIDAccepted &&
```
Can a candidate be selected as DsKKPi and DsPiKK at the same time?

vkucera commented on PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:
```diff
@@ -0,0 +1,171 @@
+
+WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
+{
+  WorkflowSpec workflow{};
+  workflow.push_back(adaptAnalysisTask<HfDsTokkpiCandidateSelector>(cfgc));
+  return workflow;
```
```suggestion
  return WorkflowSpec{adaptAnalysisTask<HfDsTokkpiCandidateSelector>(cfgc)};
```

vkucera commented on PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:
```diff
@@ -0,0 +1,171 @@
+using namespace o2::analysis::hf_cuts_ds_tokkpi;
+
+/// Struct for applying Ds to KKpi selection cuts
+struct HfDsTokkpiCandidateSelector {
```
```suggestion
struct HfDsToKKPiCandidateSelector {
```

vkucera commented on PWGHF/TableProducer/HFDsToKKPiCandidateSelector.cxx:
```diff
@@ -0,0 +1,171 @@
+  Configurable<double> pTCandMin{"pTCandMin", 1., "Lower bound of candidate pT"};
+  Configurable<double> pTCandMax{"pTCandMax", 36., "Upper bound of candidate pT"};
+  // TPC
+  Configurable<bool> requireTPC{"requireTPC", true, "Flag to require a positive Number of found clusters in TPC"};
```
`requireTPC` is not used

stefanopolitano replied:
Indeed, it is not needed inside the selector as the PID from TPC is already required in the trackselection step if I am not wrong (maybe @fgrosa can comment on this). However, the same line is present in the Dplus (https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/HFDplusToPiKPiCandidateSelector.cxx#L36) and Lc->K0sP (https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/HFLcK0sPCandidateSelector.cxx#L72) candidate selector and not used. Do you want me to remove this also from there?

fgrosa replied:
Ciao @stefanopolitano, indeed I would remove it from here: in the `HFTrackIndexSkimsCreator` the quality selection on tracks is applied, so there is no need to add other requirements here (or at least nTPCcls > 0 is not a requirement we would need anyway). You can remove it I would say, thanks!

vkucera commented on PWGHF/D2H/Tasks/taskDs.cxx:
```diff
@@ -0,0 +1,215 @@
+using namespace o2::aod::hf_cand;
+using namespace o2::aod::hf_cand_prong3;
+
+#include "Framework/runDataProcessing.h"
```
Move up

vkucera commented on PWGHF/D2H/Tasks/taskDs.cxx:
```diff
@@ -0,0 +1,215 @@
+using namespace o2;
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+using namespace o2::aod::hf_cand;
```
`aod::hf_cand` should not be needed after merging https://github.com/AliceO2Group/O2Physics/pull/966

vkucera commented on PWGHF/D2H/Tasks/taskDs.cxx:
```diff
@@ -0,0 +1,215 @@
+    registry.add("hPtGenSig", "MC particles (matched);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}}});
+    registry.add("hPtGenPrompt", "MC particles (matched, prompt);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}}});
+    registry.add("hPtGenNonPrompt", "MC particles (matched, non-prompt);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}}});
+    registry.add("hPtvsYRecSigRecoPID", "3-prong candidates (RecoPID - matched);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYRecSigPromptRecoPID", "3-prong candidates (RecoPID - matched, prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYRecSigNonPromptRecoPID", "3-prong candidates (RecoPID - matched, non-prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYRecSigRecoTopol", "3-prong candidates (RecoTopol - matched);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYRecSigPromptRecoTopol", "3-prong candidates (RecoTopol - matched, prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYRecSigNonPromptRecoTopol", "3-prong candidates (RecoTopol - matched, non-prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYRecSigRecoSkim", "3-prong candidates (RecoSkim - matched);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYRecSigPromptRecoSkim", "3-prong candidates (RecoSkim - matched, prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYRecSigNonPromptRecoSkim", "3-prong candidates (RecoSkim - matched, non-prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYGen", "MC particles (matched);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYGenPrompt", "MC particles (matched, prompt);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYGenNonPrompt", "MC particles (matched, non-prompt);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
```
```suggestion
    registry.add("hPtVsYRecSigRecoPID", "3-prong candidates (RecoPID - matched);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigPromptRecoPID", "3-prong candidates (RecoPID - matched, prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigNonPromptRecoPID", "3-prong candidates (RecoPID - matched, non-prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigRecoTopol", "3-prong candidates (RecoTopol - matched);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigPromptRecoTopol", "3-prong candidates (RecoTopol - matched, prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigNonPromptRecoTopol", "3-prong candidates (RecoTopol - matched, non-prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigRecoSkim", "3-prong candidates (RecoSkim - matched);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigPromptRecoSkim", "3-prong candidates (RecoSkim - matched, prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYRecSigNonPromptRecoSkim", "3-prong candidates (RecoSkim - matched, non-prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYGen", "MC particles (matched);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYGenPrompt", "MC particles (matched, prompt);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
    registry.add("hPtVsYGenNonPrompt", "MC particles (matched, non-prompt);#it{p}_{T}^{gen.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
```

vkucera commented on PWGHF/D2H/Tasks/taskDs.cxx:
```diff
@@ -0,0 +1,215 @@
+    registry.add("hPtGen", "MC particles (matched);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}}});
+    registry.add("hPtGenSig", "MC particles (matched);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}}});
+    registry.add("hPtGenPrompt", "MC particles (matched, prompt);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}}});
+    registry.add("hPtGenNonPrompt", "MC particles (matched, non-prompt);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}}});
```
What appears on the x axis title? `#it{p}_{T}^{gen.} (GeV/#it{c})` or `#it{p}_{T} (GeV/#it{c})`?

stefanopolitano replied:
I get `#it{p}_{T} (GeV/#it{c})`, I also checked  and it is in line with the other tasks.

vkucera commented on PWGHF/D2H/Tasks/taskDs.cxx:
```diff
@@ -0,0 +1,215 @@
+    registry.add("hPtGenSig", "MC particles (matched);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}}});
+    registry.add("hPtGenPrompt", "MC particles (matched, prompt);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}}});
+    registry.add("hPtGenNonPrompt", "MC particles (matched, non-prompt);#it{p}_{T}^{gen.} (GeV/#it{c});entries", {HistType::kTH1F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}}});
+    registry.add("hPtvsYRecSigRecoPID", "3-prong candidates (RecoPID - matched);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
+    registry.add("hPtvsYRecSigPromptRecoPID", "3-prong candidates (RecoPID - matched, prompt);#it{p}_{T}^{rec.}; #it{y}", {HistType::kTH2F, {{vbins, "#it{p}_{T} (GeV/#it{c})"}, {100, -5., 5.}}});
```
You could define an axis for *y* to avoid repetition of `{100, -5., 5.}`.

vkucera commented on PWGHF/D2H/Tasks/taskDs.cxx:
```diff
@@ -0,0 +1,215 @@
+
+  Partition<soa::Join<aod::HfCandProng3, aod::HFSelDsToKKPiCandidate, aod::HfCandProng3MCRec>> recoFlagDsCandidates = aod::hf_selcandidate_ds::isSelDsToKKPi >= selectionFlagDs || aod::hf_selcandidate_ds::isSelDsTPiKK >= selectionFlagDs;
+
+  void processMC(soa::Join<aod::HfCandProng3, aod::HFSelDsToKKPiCandidate, aod::HfCandProng3MCRec> const& candidates,
+                 soa::Join<aod::McParticles, aod::HfCandProng3MCGen> const& particlesMC, aod::BigTracksMC const& tracks)
```
`tracks` are not used
```suggestion
  void processMC(soa::Join<aod::HfCandProng3, aod::HFSelDsToKKPiCandidate, aod::HfCandProng3MCRec> const& candidates,
                 soa::Join<aod::McParticles, aod::HfCandProng3MCGen> const& particlesMC, aod::BigTracksMC const&)
```

vkucera commented on PWGHF/DataModel/HFCandidateSelectionTables.h:
```diff
@@ -93,6 +93,14 @@ DECLARE_SOA_COLUMN(IsSelDplusToPiKPi, isSelDplusToPiKPi, int); //!
 
+namespace hf_selcandidate_ds
+{
+DECLARE_SOA_COLUMN(IsSelDsToKKPi, isSelDsToKKPi, int); //!
+DECLARE_SOA_COLUMN(IsSelDsToPiKK, isSelDsToPiKK, int); //!
+} // namespace hf_selcandidate_ds
+DECLARE_SOA_TABLE(HFSelDsToKKPiCandidate, "AOD", "HFSELDSCAND", //!
```
Add documentation comments after `//!`

vkucera commented on PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:
```diff
@@ -271,6 +280,14 @@ struct HFCandidateCreator3ProngExpressions {
+      // Ds± → K± K∓ π±
+      if (flag == 0) {
+        // Printf("Checking Ds± → K± K∓ π±");
+        if (RecoDecay::isMatchedMCGen(particlesMC, particle, pdg::Code::kDs, array{+kKPlus, -kKPlus, +kPiPlus}, true, &sign)) {
```
Same comment as for `getMatchedMCRec`.

stefanopolitano commented on PWGHF/TableProducer/HFCandidateCreator3Prong.cxx:
```diff
@@ -211,6 +211,15 @@ struct HFCandidateCreator3ProngExpressions {
+      // Ds± → K± K∓ π±
+      if (flag == 0) {
+        // Printf("Checking Ds± → K± K∓ π±");
+        indexRec = RecoDecay::getMatchedMCRec(particlesMC, arrayDaughters, pdg::Code::kDs, array{+kKPlus, -kKPlus, +kPiPlus}, true, &sign);
```
If I am not wrong, that should be fine because the decay vertex of the phi meson is not reconstructed since it does not travel much from the Ds decay vertex. In addition, I am going to implement a selection on the invariant mass of the kaon couple in order to ensure that the kaons come from a phi decay. Maybe also @fgrosa can comment on this.

vkucera replied:
We are talking about MC matching here, not reconstruction. If the kaons come from a φ here, the Ds will be labelled as unmatched.

fgrosa replied:
Hi @vkucera shouldn't the method look also for resonant decays? If not then it should be modified because we also have resonant decays in the Lc -> pKpi and D+ -> Kpipi for example, that should be considered when asking for the KKpi in the final state. 
For this specific case, I think that as long as the cut on the KK mass is not added at reco level (that I would suggest to add in a second step), all the Ds to KKpi should be considered in the MC (but keep in mind that in some MC productions we might have only Ds --> phi pi --> KKpi).

fgrosa replied:
Ah indeed! It should be at least 2

fgrosa replied:
Actually, there is the same problem for D+ and Xic+ in this file. @stefanopolitano could you please fix them as well? Thanks a lot!

fgrosa commented on PWGHF/DataModel/HFCandidateSelectionTables.h:
```diff
@@ -95,10 +95,10 @@ DECLARE_SOA_TABLE(HFSelDplusToPiKPiCandidate, "AOD", "HFSELDPLUSCAND", //!
-DECLARE_SOA_COLUMN(IsSelDsToKKPi, isSelDsToKKPi, int); //!
-DECLARE_SOA_COLUMN(IsSelDsToPiKK, isSelDsToPiKK, int); //!
+DECLARE_SOA_COLUMN(IsSelDsToKKPi, isSelDsToKKPi, int); // Ds candidates in KKpi decay channel column
+DECLARE_SOA_COLUMN(IsSelDsToPiKK, isSelDsToPiKK, int); // Ds candidates in piKK decay channel column
```
The exclamation mark in `ROOT` indicate that a data member is transient (see https://root.cern.ch/root/html534/guides/users-guide/InputOutput.html), so can it be the same here? If this is the case, it is important to keep them not only as a comment

vkucera replied:
I checked and it seems only comments with `//!` go into the generated table/column documentation.

stefanopolitano replied:
Ok, so I just leave the `\\!` as it was before, right?

vkucera replied:
Yes

stefanopolitano replied:
Ok, thanks!

vkucera commented on PWGHF/TableProducer/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -879,10 +879,10 @@ struct HfTrackIndexSkimsCreator {
-     {"hMassLcToPKPi", "#Lambda_{c} candidates;inv. mass (p K #pi) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 5.}}}},
-     {"hMassDsToPiKK", "D_{s} candidates;inv. mass (K K #pi) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 5.}}}},
-     {"hMassXicToPKPi", "#Xi_{c} candidates;inv. mass (p K #pi) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 5.}}}}}};
+     {"hmassDPlusToPiKPi", "D^{#plus} candidates;inv. mass (#pi K #pi) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 5.}}}},
+     {"hmassLcToPKPi", "#Lambda_{c} candidates;inv. mass (p K #pi) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 5.}}}},
+     {"hmassDsToKKPi", "D_{s} candidates;inv. mass (K K #pi) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 5.}}}},
+     {"hmassXicToPKPi", "#Xi_{c} candidates;inv. mass (p K #pi) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 0., 5.}}}}}};
```
Do not rename these.

vkucera reviewed (dismissed):
Thanks for addressing the comments. The current version looks fine to me.

alibuild commented:
Error while checking build/O2Physics/o2 for 8758008e171a605945b6209498620a0a93586e75 at 2022-08-16 16:43:
```
## sw/BUILD/O2Physics-latest/log
/sw/SOURCES/O2Physics/941-slc7_x86-64/0/EventFiltering/PWGHF/HFFilter.cxx:1133:76: error: 'DsToPiKK' is not a member of 'o2::aod::hf_cand_prong3::DecayType'
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/941/8758008e171a605945b6209498620a0a93586e75/build_O2Physics_o2/pretty.html).

lvermunt commented:
Hi @stefanopolitano

I had a look at your Ds code. For most I agree with @vkucera that it is ready for merging besides the error highlighted by alibuild. However I have also a doubt about how you use the RecoDecay::getMatchedMCRec() function for the Ds. And also on your modification for the Dplus one.

Now I might be wrong since I just had a quick look (@fgrosa, maybe you can comment as well), but it seems to me you tag all resonant Ds -> X -> KKpi, and not just Ds -> phi pi -> KKpi as we did for the Run 2 analyses. Is this intended? If so, I would say you need to tag the different resonant channels for possibly different efficiency corrections (as currently already done for Lc -> pKpi).

For the Dplus, your modification of the depthMax to 2 is correct to also take the resonant K*(892)0 channel into account (as we did in Run 2). However, since the pdg code of the K* is not specified, I guess it also tags the D+ -> K+ K0s -> K+ pi pi channel. And this latter channel you don't want I would say.

Finally, your call to RecoDecay::getMother() in taskDs.cxx should also get the depthMax = 2 right?

Cheers, Luuk

lvermunt reviewed (commented):
See https://github.com/AliceO2Group/O2Physics/pull/941#issuecomment-1222090701 for my review.

vkucera commented:
Hi @lvermunt 

> Finally, your call to RecoDecay::getMother() in taskDs.cxx should also get the depthMax = 2 right?

`depthMax = -1` by default in `getMother()` so it does not make a difference for matched candidates.

vkucera commented:
Hi @stefanopolitano , can you fix the error?

@lvermunt @xinyepeng Do you have any remaining comments?

lvermunt commented:
Hi @vkucera, besides my two doubts on how the RecoDecay::getMatchedMCRec() function is used here which are not yet answered, and the error highlighted by alibuild, I don't have further comments.

Btw, if I understood correctly from a meeting earlier today, I think that Stefano is on holiday

xinyepeng commented:
Hi @vkucera, I  agree with @lvermunt , nothing more from my side

stefanopolitano commented:
Hi all, 
sorry for the late reply. I am addressing the points raised by @lvermunt. In particular:
- regarding the Ds I implemented a selection on the invariant mass f the KK couple (as was done in Run 2) to reject the candidates that do not include the phi decay. 
- concerning the Dplus I am not really sure how to implement this. Would it be better to look for the K0s PDG code for one of the intermediate particles of the decay chain in the HFCandidateCreator3Prong and reject all the candidates that present such code, or to implement this kind of selection in the HFDplusToPiKPiCandidateSelector?

fgrosa commented:
> > Hi all, sorry for the late reply. I am addressing the points raised by @lvermunt. In particular:
> > ```
> > * regarding the Ds I implemented a selection on the invariant mass f the KK couple (as was done in Run 2) to reject the candidates that do not include the phi decay.
> > 
> > * concerning the Dplus I am not really sure how to implement this. Would it be better to look for the K0s PDG code for one of the intermediate particles of the decay chain in the HFCandidateCreator3Prong and reject all the candidates that present such code, or to implement this kind of selection in the HFDplusToPiKPiCandidateSelector?
> > ```
> 
> On the first point I agree that the current version of the code is consistent, though in a second PR we should include the selection on the massKK and consequently the selection of the phi resonance only. Regarding the second one I agree with @lvermunt, however this requires to modify the `RecoDecay` class to pass one or more "desired" intermediate states and I would also do it in a separated PR.
Sorry for the double comment: in the D+ case there should not be issues because the D+ -> K0SK+ would have different signs (K+pi-pi+ instead of K-pi+pi+), so there is no issue.

lvermunt commented:
Hi @stefanopolitano, @fgrosa,

Thanks for your replies. 

I don't really follow your reply on the KK invariant mass. This is of course a good selection to include in the analysis. However, I was speaking about the tagging of the MC candidates with the getMatchedMCRec() function for the efficiency correction. This should always check the correct decay channel (or tag the various resonant ones), and not depend on a KK invariant mass selection being enabled. I think Fabrizio says the same. Anyhow, fine for me to address this in a 2nd PR.

Regarding the D+, you are of course right @fgrosa. Sorry for introducing noise there.

This PR is for me thus good to be merged.

Cheers, Luuk

fgrosa commented:
Hi @lvermunt, 

> Hi @stefanopolitano, @fgrosa,
> 
> Thanks for your replies.
> 
> I don't really follow your reply on the KK invariant mass. This is of course a good selection to include in the analysis. However, I was speaking about the tagging of the MC candidates with the getMatchedMCRec() function for the efficiency correction. This should always check the correct decay channel (or tag the various resonant ones), and not depend on a KK invariant mass selection being enabled. I think Fabrizio says the same. Anyhow, fine for me to address this in a 2nd PR.
> 

The massKK selection selects the Ds decaying into phi, without that selection we are not measuring Ds -> phi pi -> KK pi, but Ds -> X -> KKpi, where X is any intermediate state. Hence, as long as we don't have that selection the code is consistent without requiring the phi (but I agree that the immediately next step is to add it).

> Regarding the D+, you are of course right @fgrosa. Sorry for introducing noise there.
> 
> This PR is for me thus good to be merged.
> 
> Cheers, Luuk

lvermunt commented:
> Hi @lvermunt,
> 
> The massKK selection selects the Ds decaying into phi, without that selection we are not measuring Ds -> phi pi -> KK pi, but Ds -> X -> KKpi, where X is any intermediate state. Hence, as long as we don't have that selection the code is consistent without requiring the phi (but I agree that the immediately next step is to add it).

Sure, this I understand, and I agree the code is consistent enough (even though I'm also aware of your study for your PhD thesis that even with this massKK selection one still has a contamination of several percent by Ds -> f0(980)pi -> KKpi). My point was only that one shouldn't depend for the correct MC tagging on a selection that can be loosened or turned off. So that should be addressed in the future.

I approved and enabled the automatic merging. Unclear why the last pending check takes so long though

vkucera commented:
@jgrosseo @ktf The errors in the macOS tests are unrelated. Can you please force the merge?

jgrosseo commented:
You should be able to merge, macOS is not required.

jgrosseo commented:
Maybe you miss the approval for `EventFiltering` ?

vkucera commented:
> Maybe you miss the approval for `EventFiltering` ?

Indeed
